File: bridge/model/schedhook.go
package model

import (
	"bridge/utils"
	"fmt"
)

type SchedEvent struct {
	Prev          TaskKey `json:"prev"`
	Next          TaskKey `json:"next"`
	CommPrev      string  `json:"commPrev"`
	CommNext      string  `json:"commNext"`
	Timestamp     uint64  `json:"timestamp"`
	CPU           int32   `json:"cpu"`
	PrevIsKthread bool    `json:"prevIsKthread"`
	NextIsKthread bool    `json:"nextIsKthread"`
}

func ReadSchedEvent(parser utils.ByteParser) (SchedEvent, error) {
	var data SchedEvent

	data.Prev, _ = ReadTaskKey(parser)
	data.Next, _ = ReadTaskKey(parser)
	data.CommPrev = parser.ReadString(16)
	data.CommNext = parser.ReadString(16)
	data.Timestamp = parser.ReadUint64()
	data.CPU = parser.ReadInt32()
	data.PrevIsKthread = parser.ReadBool()
	data.NextIsKthread = parser.ReadBool()

	return data, parser.Error()
}

type SchedCap struct {
	Events []SchedEvent `json:"events"`
}

type WebsocketSchedEvent struct {
	Prev          WebsocketTaskKey `json:"prev"`
	Next          WebsocketTaskKey `json:"next"`
	CommPrev      string           `json:"commPrev"`
	CommNext      string           `json:"commNext"`
	Timestamp     string           `json:"timestamp"`
	CPU           int32            `json:"cpu"`
	PrevIsKthread bool             `json:"prevIsKthread"`
	NextIsKthread bool             `json:"nextIsKthread"`
}

type WebsocketSchedCap struct {
	Events []WebsocketSchedEvent `json:"events"`
}

func ToWebsocketSchedEvent(event SchedEvent) WebsocketSchedEvent {
	return WebsocketSchedEvent{
		Prev:          ToWebsocketTaskKey(event.Prev),
		Next:          ToWebsocketTaskKey(event.Next),
		CommPrev:      event.CommPrev,
		CommNext:      event.CommNext,
		Timestamp:     fmt.Sprintf("%d", event.Timestamp),
		CPU:           event.CPU,
		PrevIsKthread: event.PrevIsKthread,
		NextIsKthread: event.NextIsKthread,
	}
}

File: webapp/src/types/ws/schedhook.ts
import type { WebsocketTaskKey } from "./shared";
import { isWebsocketTaskKey } from "./shared";

export type WebsocketSchedEvent = {
  prev: WebsocketTaskKey;
  next: WebsocketTaskKey;
  commPrev: string;
  commNext: string;
  timestamp: string;
  cpu: number;
  prevIsKthread: boolean;
  nextIsKthread: boolean;
};

export function isSchedEvent(obj: any): obj is WebsocketSchedEvent {
  return (
    obj !== null &&
    typeof obj === "object" &&
    isWebsocketTaskKey(obj.prev) &&
    isWebsocketTaskKey(obj.next) &&
    typeof obj.commPrev === "string" &&
    typeof obj.commNext === "string" &&
    typeof obj.timestamp === "string" &&
    typeof obj.cpu === "number" &&
    typeof obj.prevIsKthread === "boolean" &&
    typeof obj.nextIsKthread === "boolean"
  );
}

export type WebsocketSchedCap = {
  events: WebsocketSchedEvent[];
};

export function isSchedCap(obj: any): obj is WebsocketSchedCap {
  return (
    obj !== null &&
    typeof obj === "object" &&
    Array.isArray(obj.events) &&
    obj.events.every((event: any) => isSchedEvent(event))
  );
}
